File: gyroball/Storage/SessionStore.swift
import Foundation
import Combine

/// SQLite-backed store for workout sessions, plus the aggregates the UI shows.
final class SessionStore: ObservableObject {

    @Published private(set) var sessions: [WorkoutSession] = []

    private let db: Database?

    init() {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory,
                                           in: .userDomainMask)[0]
            .appendingPathComponent("Gyroball", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

        db = Database(path: dir.appendingPathComponent("gyroball.sqlite").path)
        db?.run("""
            CREATE TABLE IF NOT EXISTS sessions(
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                started_at REAL NOT NULL,
                duration REAL NOT NULL,
                top_rpm REAL NOT NULL,
                avg_rpm REAL NOT NULL,
                revolutions REAL NOT NULL,
                samples TEXT NOT NULL DEFAULT '[]'
            )
            """)
        db?.run("CREATE INDEX IF NOT EXISTS idx_sessions_started ON sessions(started_at)")
        reload()
    }

    // MARK: - CRUD

    func add(_ session: WorkoutSession) {
        let samplesJSON = (try? JSONEncoder().encode(session.samples))
            .flatMap { String(data: $0, encoding: .utf8) } ?? "[]"

        db?.run("""
            INSERT INTO sessions (started_at, duration, top_rpm, avg_rpm, revolutions, samples)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            bind: [session.startedAt.timeIntervalSince1970,
                   session.duration,
                   session.topRPM,
                   session.avgRPM,
                   session.revolutions,
                   samplesJSON])
        reload()
    }

    func delete(_ session: WorkoutSession) {
        db?.run("DELETE FROM sessions WHERE id = ?", bind: [session.id])
        reload()
    }

    private func reload() {
        var loaded: [WorkoutSession] = []
        db?.run("""
            SELECT id, started_at, duration, top_rpm, avg_rpm, revolutions, samples
            FROM sessions ORDER BY started_at DESC
            """) { row in
            let samples = (try? JSONDecoder().decode(
                [Double].self, from: Data(row.string(6).utf8))) ?? []
            loaded.append(WorkoutSession(
                id: row.int64(0),
                startedAt: Date(timeIntervalSince1970: row.double(1)),
                duration: row.double(2),
                topRPM: row.double(3),
                avgRPM: row.double(4),
                revolutions: row.double(5),
                samples: samples))
        }
        sessions = loaded
    }

    // MARK: - Aggregates

    var todaySessions: [WorkoutSession] {
        sessions.filter { Calendar.current.isDateInToday($0.startedAt) }
    }

    var todayRevolutions: Double { todaySessions.reduce(0) { $0 + $1.revolutions } }
    var todaySeconds: TimeInterval { todaySessions.reduce(0) { $0 + $1.duration } }

    var allTimeTopRPM: Double { sessions.map(\.topRPM).max() ?? 0 }
    var totalRevolutions: Double { sessions.reduce(0) { $0 + $1.revolutions } }
    var totalSeconds: TimeInterval { sessions.reduce(0) { $0 + $1.duration } }
    var longestSession: WorkoutSession? { sessions.max { $0.duration < $1.duration } }
    var bestSession: WorkoutSession? { sessions.max { $0.topRPM < $1.topRPM } }

    /// Revolutions per day for the last 14 days, oldest first, zero-filled.
    var dailyHistory: [(day: Date, revs: Double)] {
        let cal = Calendar.current
        let today = cal.startOfDay(for: Date())
        var byDay: [Date: Double] = [:]
        for s in sessions {
            byDay[cal.startOfDay(for: s.startedAt), default: 0] += s.revolutions
        }
        return (0..<14).reversed().compactMap { offset in
            guard let day = cal.date(byAdding: .day, value: -offset, to: today) else { return nil }
            return (day: day, revs: byDay[day] ?? 0)
        }
    }
}

File: windows/src/Gyroball.Core/Storage/SessionStore.cs
using System.Collections.ObjectModel;
using System.Text.Json;
using CommunityToolkit.Mvvm.ComponentModel;
using Microsoft.Data.Sqlite;

namespace Gyroball.Core.Storage;

/// <summary>
/// SQLite-backed store for workout sessions plus the aggregates the UI shows.
/// Schema is identical to the macOS app's SessionStore.swift, so the two ports
/// read the same database shape. Observable for WPF binding.
/// </summary>
public sealed class SessionStore : ObservableObject, IDisposable
{
    private readonly Database _db;

    public ObservableCollection<WorkoutSession> Sessions { get; } = new();

    /// <summary>Default location: %APPDATA%\Gyroball\gyroball.sqlite (parallels ~/Library/Application Support).</summary>
    public static string DefaultPath()
    {
        var dir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "Gyroball");
        Directory.CreateDirectory(dir);
        return Path.Combine(dir, "gyroball.sqlite");
    }

    public SessionStore(string? path = null)
    {
        _db = new Database(path ?? DefaultPath());
        _db.Run("""
            CREATE TABLE IF NOT EXISTS sessions(
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                started_at REAL NOT NULL,
                duration REAL NOT NULL,
                top_rpm REAL NOT NULL,
                avg_rpm REAL NOT NULL,
                revolutions REAL NOT NULL,
                samples TEXT NOT NULL DEFAULT '[]'
            )
            """);
        _db.Run("CREATE INDEX IF NOT EXISTS idx_sessions_started ON sessions(started_at)");
        Reload();
    }

    // MARK: - CRUD

    public void Add(WorkoutSession session)
    {
        var samplesJson = JsonSerializer.Serialize(session.Samples);
        _db.Run("""
            INSERT INTO sessions (started_at, duration, top_rpm, avg_rpm, revolutions, samples)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            new object[]
            {
                ToUnix(session.StartedAt),
                session.Duration,
                session.TopRpm,
                session.AvgRpm,
                session.Revolutions,
                samplesJson
            });
        Reload();
    }

    public void Delete(WorkoutSession session)
    {
        _db.Run("DELETE FROM sessions WHERE id = ?", new object[] { session.Id });
        Reload();
    }

    private void Reload()
    {
        var loaded = new List<WorkoutSession>();
        _db.Run("""
            SELECT id, started_at, duration, top_rpm, avg_rpm, revolutions, samples
            FROM sessions ORDER BY started_at DESC
            """, row: r =>
        {
            double[] samples;
            try { samples = JsonSerializer.Deserialize<double[]>(r.GetString(6)) ?? Array.Empty<double>(); }
            catch { samples = Array.Empty<double>(); }

            loaded.Add(new WorkoutSession
            {
                Id = r.GetInt64(0),
                StartedAt = FromUnix(r.GetDouble(1)),
                Duration = r.GetDouble(2),
                TopRpm = r.GetDouble(3),
                AvgRpm = r.GetDouble(4),
                Revolutions = r.GetDouble(5),
                Samples = samples
            });
        });

        Sessions.Clear();
        foreach (var s in loaded) Sessions.Add(s);
        RaiseAggregatesChanged();
    }

    // MARK: - Aggregates

    public IEnumerable<WorkoutSession> TodaySessions =>
        Sessions.Where(s => s.StartedAt.Date == DateTime.Today);

    public double TodayRevolutions => TodaySessions.Sum(s => s.Revolutions);
    public double TodaySeconds => TodaySessions.Sum(s => s.Duration);
    public int TodaySessionCount => TodaySessions.Count();

    public double AllTimeTopRpm => Sessions.Count == 0 ? 0 : Sessions.Max(s => s.TopRpm);
    public double TotalRevolutions => Sessions.Sum(s => s.Revolutions);
    public double TotalSeconds => Sessions.Sum(s => s.Duration);
    public WorkoutSession? LongestSession => Sessions.MaxBy(s => s.Duration);
    public WorkoutSession? BestSession => Sessions.MaxBy(s => s.TopRpm);

    /// <summary>Revolutions per day for the last 14 days, oldest first, zero-filled.</summary>
    public IReadOnlyList<(DateTime Day, double Revs)> DailyHistory()
    {
        var today = DateTime.Today;
        var byDay = new Dictionary<DateTime, double>();
        foreach (var s in Sessions)
        {
            var day = s.StartedAt.Date;
            byDay[day] = byDay.GetValueOrDefault(day) + s.Revolutions;
        }

        var result = new List<(DateTime, double)>(14);
        for (int offset = 13; offset >= 0; offset--)
        {
            var day = today.AddDays(-offset);
            result.Add((day, byDay.GetValueOrDefault(day)));
        }
        return result;
    }

    private void RaiseAggregatesChanged()
    {
        foreach (var name in new[]
        {
            nameof(TodayRevolutions), nameof(TodaySeconds), nameof(TodaySessionCount),
            nameof(AllTimeTopRpm), nameof(TotalRevolutions), nameof(TotalSeconds),
            nameof(LongestSession), nameof(BestSession)
        })
        {
            OnPropertyChanged(name);
        }
    }

    private static double ToUnix(DateTime dt) =>
        new DateTimeOffset(dt.ToUniversalTime()).ToUnixTimeMilliseconds() / 1000.0;

    private static DateTime FromUnix(double seconds) =>
        DateTimeOffset.FromUnixTimeMilliseconds((long)(seconds * 1000)).LocalDateTime;

    public void Dispose() => _db.Dispose();
}
